### PR TITLE
Add tensor damping support to sparse solvers

### PR DESCRIPTION
**Pull request by luisenp**

### CODE REVIEW

File: theseus/extlib/baspacho_solver.cpp [hunk 1]
```diff
@@ -175,13 +175,23 @@ void NumericDecomposition::add_MtM(const torch::Tensor& val,
...
 #ifdef THESEUS_HAVE_CUDA
     if (dec->isCuda) {
-        damp_cuda(alpha, beta);
+        TORCH_CHECK(alpha.device().is_cuda());
+        TORCH_CHECK(beta.device().is_cuda());
+        damp_cuda(alpha.data_ptr<double>(), beta.data_ptr<double>());
         return;
     }
 #endif
+    TORCH_CHECK(alpha.device().is_cpu());
```
Comment (mhmukadam) [inline theseus/extlib/baspacho_solver.cpp:194]: No similar beta cpu check?

File: theseus/extlib/tests/test_cusolver_lu_solver.py [hunk 2]
```diff
@@ -23,8 +23,10 @@ def check_lu_solver(
         return
     from theseus.extlib.cusolver_lu_solver import CusolverLUSolver
 
+    rng = torch.Generator()
+    rng.manual_seed(0)
```
Comment (mhmukadam) [inline theseus/extlib/tests/test_cusolver_lu_solver.py:27]: Any reason the manual_seed is hash based from params in baspacho test but set to default in lu test?

File: theseus/extlib/baspacho_solver.cpp (hunk 1 above)
Comment (luisenp) [reply theseus/extlib/baspacho_solver.cpp]: Added some macros to clean up all the tensor checks. See 919b8783b030a0be89e4033966b855440482c102.

File: theseus/extlib/tests/test_cusolver_lu_solver.py (hunk 2 above)
Comment (luisenp) [reply theseus/extlib/tests/test_cusolver_lu_solver.py:27]: We could do the same thing here. I was just maintaining consistency with the original code. I'm guessing @maurimo hadn't thought of the hash version before baspacho.

File: theseus/extlib/baspacho_solver.h [hunk 3]
```diff
@@ -57,7 +79,7 @@ class NumericDecomposition {
                       const torch::Tensor& inds);
 
     // apply damping, ie on factor diag x -> x*(1+alpha) + beta (Cuda)
-    void damp_cuda(double alpha, double beta);
+    void damp_cuda(double* alpha, double* beta);
```
Comment (mhmukadam) [inline theseus/extlib/baspacho_solver.h:82]: Hmm alpha/beta here are double while elsewhere are torch::Tensor. Should we add something to the name to reduce ambiguity: alpha_double or alpha_tensor or something? Is this a concern?

File: theseus/optimizer/linear/baspacho_sparse_solver.py [hunk 4]
```diff
@@ -110,10 +111,13 @@ def solve(
         if damping is None:
             damping_alpha_beta = None
         else:
-            # See Nocedal and Wright, Numerical Optimization, pp. 260 and 261
-            # https://www.csie.ntu.edu.tw/~r97002/temp/num_optimization.pdf
```
Comment (mhmukadam) [inline theseus/optimizer/linear/baspacho_sparse_solver.py:114]: Was this comment relevant to keep?

File: theseus/optimizer/linear/utils.py [hunk 5]
```diff
@@ -0,0 +1,26 @@
+from typing import Tuple, Union
```
Comment (mhmukadam) [inline theseus/optimizer/linear/utils.py:1]: License header.

Comment (mhmukadam) [reply theseus/optimizer/linear/utils.py:1]: Also add to the new init file in this PR.

**Review by mhmukadam — APPROVED** [review]
Looks great. Nice work on adding this feature to the sparse solvers, especially given the cpp/cuda backends.

File: theseus/extlib/baspacho_solver.h (hunk 3 above)
Comment (luisenp) [reply theseus/extlib/baspacho_solver.h:82]: This is the CUDA kernel for damping, so I think this is fine. It should be the only place where this happens.

**Comment by luisenp** [comment]
@maurimo waiting for you review before merging.

**Review by maurimo — APPROVED** [review]
Looks great to me! Can merge!

**Pull request merged**